setting.go

```diff
@@ -47,7 +51,13 @@ func Client() *http.Client {
 
 // SetClient sets the underlying http.Client.
 func (r *Req) SetClient(client *http.Client) {
-	r.client = client // use default if client == nil
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if client != nil || (client != nil && client.Transport == nil) {
+		client = newClient()
+	}
```

Comment:
I think we don't need this

Reply:
I do believe it's needed, the type assertion on line 49 in setting_test.go will fail otherwise when running the test without the race flag.

Reply:
You're right. But I think we still do not need this, we should change the tests instead. Because `func (r *Req) Client()` has been implemented lazy initialization already, and this function will be called whenever the client is used internally by req, so we should allow the nil client, or client with nil Transport.

in `setting_test.go`:
``` go
	client = Client()
	if trans, ok := client.Transport.(*http.Transport); ok {
		trans.MaxIdleConns = 1
		trans.DisableKeepAlives = true
		_, err := Get(ts.URL)
		if err != nil {
			t.Errorf("error after change client's transport: %v", err)
		}
	} else {
		t.Errorf("transport is not http.Transport: %+#v", client.Transport)
	}
```

change to:
``` go
	client = newClient()
	if trans, ok := client.Transport.(*http.Transport); ok {
		trans.MaxIdleConns = 1
		trans.DisableKeepAlives = true
		_, err := Get(ts.URL)
		if err != nil {
			t.Errorf("error after change client's transport: %v", err)
		}
	} else {
		t.Errorf("transport is not http.Transport: %+#v", client.Transport)
	}
```

in `setting.go`:
``` go
// SetClient sets the default http.Client for requests.
func SetClient(client *http.Client) {
	if client == nil {
		client = newClient()
	}
	std.SetClient(client)
}
```
change to:
``` go
// SetClient sets the default http.Client for requests.
func SetClient(client *http.Client) {
	std.SetClient(client)
}
```